packages/material-ui/package.json

```diff
@@ -39,6 +39,7 @@
     "typescript": "tslint -p tsconfig.json \"{src,test}/**/*.{spec,d}.{ts,tsx}\" && tsc -p tsconfig.json && tsc -p tsconfig.build.json"
   },
   "peerDependencies": {
+    "@material-ui/styled-engine": "^5.0.0-alpha.1",
```

Comment:
That's not indicative of zero-config. Packages managers are discussing to re-introduce automatic installation of peer dependencies. But until then every peer dependency implies extra config steps.

Reply:
Should we move this then to the `dependencies`?

Reply:
Yeah. Which means the dependency alias won't work. It'll have to be either `resolutions` (`yarn` only I think) or an alias in your bundler.

Reply:
Done - https://github.com/mui-org/material-ui/pull/22435/commits/a37b64fedc8b91cdf34002e319bad1ab3da740d0 let me know if we should improve the docs further

Reply:
From what I understand, if you are a new user, and you want to use Material-UI v5, you will have 2 dependencies to install:

- @material-ui/core
- @emotion/core

The two other packages of emotion can be direct dependencies of @material-ui/styled-engine has have no singletons. Hopefully, @Andarist can confirm it :). 

- @emotion/styled
- emotion-theming

At this point, developers will often copy and paste the installation command. I think that adding a 3rd dependency (@material-ui/styled-engine) as a requirement, if it saves styled components + CRA users to eject, could be considered a viable option in this "zero-config" objective.

I don't think that it should block the pull request. We can resolve that later on. I also think that the proposal of Sebastian is better to start experimenting with because it yields a better DX by default **and** has a forcing feedback function (we will hear about it if configuring styled-components with CRA is really a frequent use case and painful).

Reply:
> At this point, developers will often copy and paste the installation command. I think that adding a 3rd dependency (@material-ui/styled-engine) as a requirement, if it saves styled components + CRA users to eject, could be considered a viable option in this "zero-config" objective.

I believe the same - this doesn't break the 0 config promise in my eyes.

In general, peerDependencies kinda suck and I wish we & you could avoid this but in this particular case, it seems like a lesser evil. It's the same situation as with React - you really want to ever have only one instance of it.

As to `emotion-theming` - we are *very* close to releasing Emotion 11 (1 issue to solve + probably some short RC period after that) and with Emotion 11 (which isn't is just an evolution and not a revolution like 9->10 was) there won't be `emotion-theming` package - those things are included in `@emotion/react` (the new name of `@emotion/core` as many people were complaining about the old, vague, name). 

`@emotion/styled` is still there though - but yes, it's much safer for it to be a regular dependency if you want that.

Reply:
I think removing zero-config should be a blocker. It makes sense to resolve minor issues later but zero-config should stay as a requirement. It wouldn't be that bad if we already had some setup required but adding one makes using this library way more complicated. You might be laughing how I consider adding one dependency is "complicated" but you have to understand that people need to know about this requirement first and then have to solve it. Not everybody reads the documentation from page 1. Especially if they're used to Material-UI. We don't need feedback for this. This new configuration will leave some behind.

> I believe the same - this doesn't break the 0 config promise in my eyes.

I'm not following how this is opinionated. Either we need this dependency or we don't. If it's optional then declare it as such in `peerDependenciesMeta`. If `yarn add @material-ui/core` is no longer enough then the current 0-config promise is broken because I have to know about the additional requirement and I have to act on it. And it gets considerably more complicated if the user has to resolve potential future major version conflicts e.g. "does /core v5 work with emotion 11 or 10?".

Reply:
From my PoV - 0 config approach is mostly about providing 0 setup code + 0 tooling configuration, not about installation steps ¯\\_(ツ)_/¯ This is also *very* easily discoverable because even if u forget to install this then your app will most likely crash when rendering one of the first components so you will get reminded about this pretty quickly.

> And it gets considerably more complicated if the user has to resolve potential future major version conflicts e.g. "does /core v5 work with emotion 11 or 10?".

This is a valid concern and I'm afraid that I don't have a food answer for that - there is always some complexity when depending on a package like this. Those problems are not exclusive to Emotion though, they are just related to this class of packages. I strongly suspect that making this a regular dependency (and you would need that to provide 0config experience from your PoV) would bring a lot more bugs that manifest themselves at random times and are very hard to debug for inexperienced users.

Reply:
I don't think it is uncommon in cases when using emotion as a style engine, to have these as peer dependencies and add them together with the UI library installation command.

![image](https://user-images.githubusercontent.com/4512430/93772801-90198a00-fc1f-11ea-975e-18452b2a2b49.png)

![image](https://user-images.githubusercontent.com/4512430/93772833-99a2f200-fc1f-11ea-85f4-046d5620fadb.png)

As these has to be only one instance of `@emotion/core` & `emotion-theming`, I don't see any other way at this moment for solving this. Moreover,  as we want to support `styled-components` as an engine as well, I think it won't be best to include them in the bundle by default. Now for using `styled-components` as an engine, yes we have additional setup (adding the alias `@material-ui/styled-engine` -> `@material-ui/styled-engine-sc`), but again this is not the default scenario.

Reply:
> I don't think it is uncommon in cases when using emotion as a style engine, to have these as peer dependencies and add them together with the UI library installation command.

I am not arguing whether this is common or not. I'm arguing against having this requirement. Something existing in the world is not an argument for its existence. It'll really hurt prototyping with our library. It was obvious to me that zero-config is a definitive requirement to stay. I think we're way to hand-wavy about this fact (same with the variants). We're adding a lot of complexity to the styling solution and so far I haven't seen how the overall story got improved.

The original plan for unstyled components was to be style-engine agnostic and now we added more configuration for the style engine. This seems like a step backwards to me.

Especially considering that Olivier argues that this issue shouldn't be a blocker but from how I understand Marjia she's already saying the problem can't be solved at all.

Reply:
@eps1lon As far as I know, emotion and styled-components host a singleton that guarantees the correct CSS injection order, a theme context, and probably more states I'm not aware of. [styled-components](https://styled-components.com/docs/faqs#i-am-a-library-author-should-i-bundle-styledcomponents-with-my-library) encourages libraries to have them as peer dependencies. Does your requirement for a single dependency maximum to install discalifies styled-components as the default engine?

@Andarist I have seen react-select have emotion as a direct dependency, so does chakra v1: https://github.com/chakra-ui/chakra-ui/issues/615#issuecomment-616141426. What should we have as a peer dependency?

From my PoV, we can ask developers to install 2-3-4 dependencies if we have good reasons. I wouldn't worry too much about it. Sure, the fewer the better but the growth of react-bootstrap (2 deps) and chakra (4 deps) are, IMHO good indicators that it's not a bootlneck for trying a library out. It's still **one** command to copy and paste. And as for long term implications, isn't peer dependency meant to dodge long term issues?

Reply:
> @Andarist I have seen react-select have emotion as a direct dependency so does chakra v1: chakra-ui/chakra-ui#615 (comment). What should we have as a peer dependency?

Interesting, wasn't aware that they use emotion as a regular dep. Gonna ask Mitchell about this.